File: src/exercises/simple_linked_list/typescript/template.ts
export class Element<T> {
  public value: T;
  public next: Element<T> | null = null;

  constructor(value: T) {
    this.value = value;
  }
}

export class SimpleLinkedList<T> {
  private headNode: Element<T> | null = null;
  private countNodes: number = 0;

  push(value: T): void {
    // Your code here
  }

  pop(): T {
    // Your code here
    throw new Error("list is empty");
  }

  get length(): number {
    return this.countNodes;
  }

  toArray(): T[] {
    // Your code here
    return [];
  }

  reverse(): SimpleLinkedList<T> {
    // Your code here
    return this;
  }
}

export function list(initialValues: number[], operations: any[]): Record<string, any> {
  // Your code here
  return {};
}

File: src/exercises/simple_linked_list/ocaml/template.ml
type 'a node = {
  value : 'a;
  mutable next : 'a node option;
}

type 'a simple_linked_list = {
  mutable head : 'a node option;
  mutable size : int;
}

let create () = { head = None; size = 0 }

let push list v =
  (* Your code here *)
  ()

let pop list =
  (* Your code here *)
  None

let list_ops (initial_values : int list) (operations : (string * int option) list) : bool =
  (* Your code here *)
  true
